File: catalogue-api/src/test/java/io/quartic/catalogue/api/DatasetLocatorTests.java
package io.quartic.catalogue.api;

import org.junit.Test;

import java.io.IOException;

import static io.quartic.common.serdes.ObjectMappersKt.objectMapper;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;

public abstract class DatasetLocatorTests<T extends DatasetLocator> {
    protected abstract T locator();
    protected abstract String json();

    @Test
    public void deserialize_as_expected() throws IOException {
        T datasetLocator = objectMapper().readValue(json(), (Class<T>) DatasetLocator.class);
        assertThat(datasetLocator, equalTo(locator()));
    }
}

File: catalogue-api/src/test/kotlin/io/quartic/catalogue/api/DatasetLocatorTests.kt
package io.quartic.catalogue.api

import io.quartic.common.serdes.OBJECT_MAPPER
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Test

abstract class DatasetLocatorTests<out T : DatasetLocator> {
    protected abstract fun locator(): T
    protected abstract fun json(): String

    @Test
    fun deserialize_as_expected() {
        val datasetLocator = OBJECT_MAPPER.readValue(json(), DatasetLocator::class.java as Class<T>)
        assertThat(datasetLocator, equalTo(locator()))
    }
}
